internal/pkg/cli/pipeline_delete.go

```diff
@@ -0,0 +1,228 @@
+)
+
+var (
+	errNoPipelineInWorkspace = errors.New("there was no pipeline manifest found in your workspace. Please run `ecs-preview pipeline init` to create an pipeline")
+
+	errPipelineDeleteCancelled = errors.New("pipeline delete cancelled - no changes made")
+)
+
+type deletePipelineVars struct {
+	*GlobalOpts
+	SkipConfirmation bool
+	DeleteSecret     bool
+	PipelineName     string // Might not need to pass in as flag?
+	PipelineSecret   string // Might not need to pass in as flag?
+}
+
+type deletePipelineOpts struct {
+	deletePipelineVars
+
+	// Interfaces to dependencies
+	pipelineDeployer pipelineDeployer
+	prog             progress
+	secretsmanager   archer.SecretsManager
+	ws               archer.Workspace
+}
+
+func newDeletePipelineOpts(vars deletePipelineVars) (*deletePipelineOpts, error) {
+	ws, err := workspace.New()
+	if err != nil {
+		return nil, fmt.Errorf("workspace cannot be created: %w", err)
+	}
+
+	data, err := ws.ReadFile(workspace.PipelineFileName)
+	if err != nil {
+		return nil, fmt.Errorf("read pipeline file %s: %w", workspace.PipelineFileName, err)
+	}
+	pipeline, err := manifest.UnmarshalPipeline(data)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal pipeline file %s: %w", workspace.PipelineFileName, err)
+	}
+
+	secretsmanager, err := secretsmanager.NewStore()
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create secrets manager: %w", err)
+	}
+
+	p := session.NewProvider()
+	defaultSession, err := p.Default()
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &deletePipelineOpts{
+		deletePipelineVars: vars,
+		prog:               termprogress.NewSpinner(),
+		secretsmanager:     secretsmanager,
+		pipelineDeployer:   cloudformation.New(defaultSession),
+		ws:                 ws,
+	}
+	opts.PipelineName = pipeline.Name
+	if secret, ok := (pipeline.Source.Properties["access_token_secret"]).(string); ok {
+		opts.PipelineSecret = secret
+	}
+
+	return opts, nil
+}
+
+// Ask prompts for fields that are required but not passed in.
+func (o *deletePipelineOpts) Ask() error {
+	if o.SkipConfirmation {
+		return nil
+	}
+
+	deleteConfirmed, err := o.prompt.Confirm(
+		fmt.Sprintf(pipelineDeleteConfirmPrompt, o.PipelineName, o.ProjectName()),
+		pipelineDeleteConfirmHelp)
+
+	if err != nil {
+		return fmt.Errorf("pipeline delete confirmation prompt: %w", err)
+	}
+
+	if !deleteConfirmed {
+		return errPipelineDeleteCancelled
+	}
+
+	return nil
+}
+
+// Validate returns an error if the flag values passed by the user are invalid.
+func (o *deletePipelineOpts) Validate() error {
+	if o.ProjectName() == "" {
+		return errNoProjectInWorkspace
+	}
+
+	if o.PipelineName == "" {
+		return errNoPipelineInWorkspace
+	}
+
+	return nil
+}
+
+// Execute deletes the secret and pipeline stack
+func (o *deletePipelineOpts) Execute() error {
+	if err := o.deleteSecret(); err != nil {
+		return err
+	}
+
+	if err := o.deleteStack(); err != nil {
+		return err
+	}
```

Comment:
Can we ignore the errors if the secret and stack are already deleted? 

Maybe we can write something like `log.Successf("Secret %s is already removed.", o.PipelineSecret)`

Reply:
Oh, is this an error we handle like this elsewhere? I was mostly following suit of what  [`app_delete.go`](https://github.com/aws/amazon-ecs-cli-v2/blob/master/internal/pkg/cli/app_delete.go#L116-L117) does for the stack deletion. On the underlying [`cloudformation.DeleteStack`](https://github.com/aws/amazon-ecs-cli-v2/blob/1a91c0ea9a5025e6b9ff85a109cee06efb8350b4/internal/pkg/deploy/cloudformation/stack.go#L14-L22) method we already wrap some errors. Wondering if you think we should add some sugar on top of that?

For the deletion of secret, the errors returned by [secretsmanager.DeleteSecret](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DeleteSecret.html) don't seem to return like, an idempotency error. Unless I'm misunderstanding what you mean when you say these things are already deleted?

Reply:
`app_delete` calls `delete-stack` API which does not return an error if stack does not exist. However `describe-stacks` will return `ValidationError` if stack does not exist. Maybe we can use `delete-stack` in pipeline delete as well so that we don't need to validate if pipeline already exists?

Reply:
Okay so, with the latest changes (see: 3b9bf4c) here's what's happening with the following workflow:
1.  Run`pipeline init` - creates new secret and writes pipeline.yml
2. Run`pipeline update` - deploys pipeline via CFN 
3.  From the  AWS Console, I schedule the deletion of the secret I just made through `pipeline init`. This schedules the secret for deletion in N days.
4. run `pipeline delete` :
```
~/aws-reinvent-2019-trivia-backend(master)$ ecs-preview pipeline delete


? Are you sure you want to delete pipeline pipeline-backend-SoManyHs-aws-reinvent-trivia-2019-backend from project backend? Yes


? Are you sure you want to delete the source secret github-token-backend-aws-reinvent-trivia-2019-backend associated with pipeline pipeline-backend-SoManyHs-aws-reinvent-trivia-2019-backend? Yes
✔ Deleted secret github-token-backend-aws-reinvent-trivia-2019-backend.
✔ Deleted pipeline pipeline-backend-SoManyHs-aws-reinvent-trivia-2019-backend from project backend.

~/aws-reinvent-2019-trivia-backend(master)$ aws secretsmanager list-secrets --region us-east-2
{
    "SecretList": [
    ]
}
```
So as far as idempotency for deleting the secret, it seems to be okay for both deleting the secret and the stack (since the stack only refers to the name of the secret, and does not actually create it so there wouldn't be any stack drift).

Reply:
> `app_delete` calls `delete-stack` API which does not return an error if stack does not exist. However `describe-stacks` will return `ValidationError` if stack does not exist. Maybe we can use `delete-stack` in pipeline delete as well so that we don't need to validate if pipeline already exists?

I'm not seeing this in [app_delete](https://github.com/aws/amazon-ecs-cli-v2/blob/master/internal/pkg/cli/app_delete.go#L180) or in [env_delete](https://github.com/aws/amazon-ecs-cli-v2/blob/5fdd850c05ee2c7dd8cecc27bc96e2db293f6e4f/internal/pkg/deploy/cloudformation/env.go#L45) -- am I missing something?

Reply:
Here:

https://github.com/aws/amazon-ecs-cli-v2/blob/5db6427337f08f7ac9e5ba0b2f38f324a085f53e/internal/pkg/cli/app_delete.go#L192


https://github.com/aws/amazon-ecs-cli-v2/blob/1a91c0ea9a5025e6b9ff85a109cee06efb8350b4/internal/pkg/deploy/cloudformation/stack.go#L14-L19